### PR TITLE
Display Statistics Table on Assessment Overview on Course Statistics

### PR DESCRIPTION
**Pull request by bivanalhar**
## Feature

We implemented the Assessments Statistics Table, which displays the overview of each assessments within a course, and practically how students inside the course perform generally inside that course. We provide the following information, for each assessment:

- Under which Tab and Category the assessment belongs to
- Statistics regarding the number of students who submitted successfully, submitted late, and still attempting
- The average and standard deviation for both students' grades and also time taken for them to finish the assessment

## Overview

Screenshot 2024-03-15 at 3 38 16 PM

### CODE REVIEW

File: app/controllers/concerns/course/statistics/reference_times_concern.rb [hunk 1]
```diff
@@ -15,6 +15,7 @@ def personal_end_at_hash(assessment_id_array, course_id)
         ) clpi
         ON cpt.lesson_plan_item_id = clpi.id
       ),
+
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/controllers/concerns/course/statistics/reference_times_concern.rb:18]: oh this is just for the sake of readability only..

File: client/app/lib/helpers/url-helpers.js [hunk 2]
```diff
@@ -136,12 +136,20 @@ function getCurrentPath() {
   return match && match[1];
 }
 
+function getLastPartOfCurrentPath() {
+  const currentPath = getCurrentPath();
+  const currentPathParts = currentPath.split('/');
+
+  return currentPathParts[currentPathParts.length - 1];
+}
+
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/lib/helpers/url-helpers.js]: done

File: client/app/bundles/course/statistics/pages/StatisticsIndex/index.tsx [hunk 3]
```diff
@@ -61,19 +65,30 @@ const allTabs = {
...
     allTabs.courseTab,
+    allTabs.assessmentTab,
   ];
 
+  const lastPartOfCurrentPath = getLastPartOfCurrentPath();
+  const defaultTabIndex = tabs.findIndex(
+    (tab) => tab.href === lastPartOfCurrentPath,
+  );
+
+  const [tabValue, setTabValue] = useState(defaultTabIndex);
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/statistics/pages/StatisticsIndex/index.tsx]: done

File: app/controllers/concerns/course/statistics/counts_concern.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    submitted_submissions_count = submitted_submissions.each_with_object(Hash.new(0)) do |assessment_id, counts|
+      counts[assessment_id] += 1
+    end
+
+    submitted_submissions_count.to_h { |assessment_id, count| [assessment_id, count] }
+  end
+
+  def num_late_students_hash
+    @personal_end_at_hash = personal_end_at_hash(@assessments.pluck(:id), current_course.id)
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply app/controllers/concerns/course/statistics/counts_concern.rb:41]: done.. now each of the function runs around 2-4s for the course with personal times

**Review by cysjonathan — CHANGES_REQUESTED** [review]
1. Add a column "Starts at" and default ordering should be based on "Starts at" first, then "Title" if there is a tie.
2. Center the # Attempted/Submitted/Late and make the numbers clickable, pointing to the /submissions page.
3. Avg Grade / Total Grade, Total Grade should be a whole number, remove the 2 d.p.
4. Avg Grade can limit to 1 d.p., ideally truncate to whole number if it is x.0
5. Avg Time / Stdev Time ideally should truncate to largest significant value (e.g. 42 days or 2 days or 17 hours or 3 minutes)

**Comment by bivanalhar** [comment]
> 1. Add a column "Starts at" and default ordering should be based on "Starts at" first, then "Title" if there is a tie.
done

> 2. Center the # Attempted/Submitted/Late and make the numbers clickable, pointing to the /submissions page.
done

> 3. Avg Grade / Total Grade, Total Grade should be a whole number, remove the 2 d.p.
Sometimes the grades can be not a whole number, in this case the decimal points are set to either 0 or 1

> 4. Avg Grade can limit to 1 d.p., ideally truncate to whole number if it is x.0
done

> 5. Avg Time / Stdev Time ideally should truncate to largest significant value (e.g. 42 days or 2 days or 17 hours or 3 minutes)
not doing it right now

**Review by cysjonathan — APPROVED** [review]

**Pull request merged**